internals/overlord/logstate/loki/loki.go

```diff
@@ -0,0 +1,188 @@
+	if err != nil {
+		return fmt.Errorf("encoding request to JSON: %v", err)
+	}
+
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.remoteURL, bytes.NewReader(jsonReq))
+	if err != nil {
+		return fmt.Errorf("creating HTTP request: %v", err)
+	}
+	httpReq.Header.Set("Content-Type", "application/json; charset=utf-8")
+	httpReq.Header.Set("User-Agent", fmt.Sprintf("pebble/%s", cmd.Version))
+
+	resp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return err
+	}
+
+	return c.handleServerResponse(resp)
+}
+
+func (c *Client) emptyBuffer() {
+	for svc := range c.entries {
+		c.entries[svc] = c.entries[svc][:0]
+	}
+	c.numEntries = 0
+}
+
+func (c *Client) buildRequest() request {
+	// Sort keys to guarantee deterministic output
+	services := make([]string, 0, len(c.entries))
+	for svc, entries := range c.entries {
+		if len(entries) == 0 {
+			continue
+		}
+		services = append(services, svc)
+	}
+	sort.Strings(services)
+
+	var req request
+	for _, service := range services {
+		entries := c.entries[service]
+		stream := stream{
+			Labels: map[string]string{
+				"pebble_service": service,
+			},
+			Entries: entries,
+		}
+		req.Streams = append(req.Streams, stream)
+	}
+	return req
+}
+
+type request struct {
+	Streams []stream `json:"streams"`
+}
+
+type stream struct {
+	Labels  map[string]string `json:"stream"`
+	Entries []lokiEntry       `json:"values"`
+}
+
+type lokiEntry [2]string
+
+// handleServerResponse determines what to do based on the response from the
+// Loki server. 4xx and 5xx responses indicate errors, so in this case, we will
+// bubble up the error to the caller.
+func (c *Client) handleServerResponse(resp *http.Response) error {
+	defer func() {
+		// Drain request body to allow connection reuse
+		// see https://pkg.go.dev/net/http#Response.Body
+		_, _ = io.Copy(io.Discard, io.LimitReader(resp.Body, 1024*1024))
+		_ = resp.Body.Close()
+	}()
+
+	code := resp.StatusCode
+	switch {
+	case code == http.StatusTooManyRequests:
+		// For 429, don't drop logs - just retry later
+		return errFromResponse(resp)
+
+	case code >= 500:
+		// 5xx indicates a problem with the server, so don't drop logs (retry later)
+		return errFromResponse(resp)
+
+	case code >= 400:
+		// 4xx indicates a client problem, so drop the logs (retrying won't help)
+		logger.Noticef("Target %q: request failed with status %d, dropping %d logs",
+			c.targetName, code, c.numEntries)
+		c.emptyBuffer()
+		return errFromResponse(resp)
+	}
+
+	c.emptyBuffer()
+	return nil
+}
+
+// errFromResponse generates an error from a failed *http.Response.
+// NB: this function reads the response body.
+func errFromResponse(resp *http.Response) error {
+	// Read response body to get more context
+	body, err := io.ReadAll(io.LimitReader(resp.Body, 1024))
+	if err != nil {
+		return fmt.Errorf("HTTP %d error, but cannot read response: %v", resp.StatusCode, err)
+	}
+	return fmt.Errorf("HTTP %d error, response %q", resp.StatusCode, body)
```

Comment:
It doesn't seem ideal to dump 1kb of arbitrary unknown data on the user terminal or logs, unless that's a debug message that was explicitly requested as such.

Reply:
Fair point. We'll send the response body to debug logs, and just return the HTTP status in the error.